src/OrleansConsulUtils/ConsulBasedMembershipTable.cs

```diff
@@ -0,0 +1,202 @@
+        /// <returns></returns>
+        /// <remarks>
+        /// Consul Membership Provider does not support the extended Membership Protocol,
+        /// therefore there is no MembershipTable to Initialise
+        /// </remarks>
+        public Task InitializeMembershipTable(GlobalConfiguration config, Boolean tryInitTableVersion, TraceLogger traceLogger)
+        {
+            Init(config.DeploymentId, config.DataConnectionString, traceLogger);
+
+            return TaskDone.Done;
+        }
+
+        private void Init(String deploymentId, String dataConnectionString, TraceLogger traceLogger)
+        {
+            _logger = traceLogger;
+            _deploymentId = deploymentId;
+            _connectionString = dataConnectionString;
+
+            _consulClient =
+                new ConsulClient(
+                    new ConsulClientConfiguration
+                    {
+                        Address = new Uri(dataConnectionString)
+                    });
+        }
+
+        public async Task<MembershipTableData> ReadRow(SiloAddress siloAddress)
+        {
+            var siloRegistration = await GetConsulSiloRegistration(siloAddress);
+
+            return AssembleMembershipTableData(siloRegistration);
+        }
+
+        public async Task<MembershipTableData> ReadAll()
+        {
+            var deploymentKVAddresses = await _consulClient.KV.List(ConsulSiloRegistration.ParseDeploymentKVPrefix(_deploymentId));
+            if (deploymentKVAddresses.Response == null)
+            {
+                _logger.Verbose("Could not find any silo registrations for deployment {0}.", _deploymentId);
+                return new MembershipTableData(null);
+            }
+
+            var allSiloRegistrations = deploymentKVAddresses.Response.Select(siloKV => ConsulSiloRegistration.FromKVPair(_deploymentId, siloKV)).ToArray();
+
+            return AssembleMembershipTableData(allSiloRegistrations);
+        }
+
+        public async Task<Boolean> InsertRow(MembershipEntry entry, TableVersion tableVersion)
+        {
+            try
+            {
+                //Use "0" as the eTag then Consul KV CAS will treat the operation as an insert and return false if the KV already exiats.
+                var consulSiloRegistration = ConsulSiloRegistration.FromMembershipEntry(_deploymentId, entry, "0");
+                var insertKV = consulSiloRegistration.ToKVPair();
+
+                var tryUpdate = await _consulClient.KV.CAS(insertKV);
+                if (!tryUpdate.Response)
+                {
+                    _logger.Verbose("ConsulMembershipProvider failed to insert the row because a registration already exists for silo {0}.", entry.SiloAddress);
+                    return false;
+                }
+
+                return true;
+            }
+            catch (Exception ex)
+            {
+                _logger.Verbose("ConsulMembershipProvider failed to insert registration for silo {0}; {1}.", entry.SiloAddress, ex);
```

Comment:
I recommend logging this and all exception at least on Info level, maybe even warning here.